File: chesstopia-frontend/src/lib/position.ts
import type { Position, Square } from '@chesstopia/openapi-client';

export type PieceCode =
  | 'wK' | 'wQ' | 'wR' | 'wB' | 'wN' | 'wP'
  | 'bK' | 'bQ' | 'bR' | 'bB' | 'bN' | 'bP';

/** board[rank][file], rank 0 = Reihe 8, file 0 = a-Linie. */
export type Board = (PieceCode | null)[][];
export type Side = 'w' | 'b';

const RANKS = ['ONE', 'TWO', 'THREE', 'FOUR', 'FIVE', 'SIX', 'SEVEN', 'EIGHT'];
const FILES = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
const TYPE_LETTER: Record<string, string> = {
  KING: 'K', QUEEN: 'Q', ROOK: 'R', BISHOP: 'B', KNIGHT: 'N', PAWN: 'P',
};

function rankIdx(square: Square): number {
  return 7 - RANKS.indexOf(square.rank);
}
function fileIdx(square: Square): number {
  return FILES.indexOf(square.file);
}

export function toBoard(position: Position): Board {
  const board: Board = Array.from({ length: 8 }, () => Array<PieceCode | null>(8).fill(null));
  for (const { square, piece } of position.board) {
    const code = (piece.color === 'WHITE' ? 'w' : 'b') + TYPE_LETTER[piece.type];
    board[rankIdx(square)][fileIdx(square)] = code as PieceCode;
  }
  return board;
}

export function sideOf(position: Position): Side {
  return position.sideToMove === 'BLACK' ? 'b' : 'w';
}

const FILE_BY_LETTER: Record<string, Square['file']> = {
  a: 'A', b: 'B', c: 'C', d: 'D', e: 'E', f: 'F', g: 'G', h: 'H',
};
const RANK_BY_DIGIT: Record<string, Square['rank']> = {
  '1': 'ONE', '2': 'TWO', '3': 'THREE', '4': 'FOUR', '5': 'FIVE', '6': 'SIX', '7': 'SEVEN', '8': 'EIGHT',
};

/** `"e2"` → `{ file: 'E', rank: 'TWO' }`. */
export function parseSquare(name: string): Square {
  const file = FILE_BY_LETTER[name[0]];
  const rank = RANK_BY_DIGIT[name[1]];
  if (name.length !== 2 || !file || !rank) throw new RangeError(`Kein Feldname: ${name}`);
  return { file, rank };
}

File: chess-engine/src/commonMain/kotlin/io/chesstopia/engine/Position.kt
@file:OptIn(ExperimentalJsExport::class)

package io.chesstopia.engine

import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport

/**
 * Eine Stellung — das Brett und die fünf Zusatzangaben, die ein Zug fortführt.
 *
 * `board` listet nur die besetzten Felder auf, als [PlacedPiece]-Paare — dieselbe
 * Form wie der REST-Kontrakt und die Persistenz. Kein Feldindex, keine
 * Leserichtung: ein Feld hat eine Figur oder es steht nicht in der Liste.
 * `Array` statt `List` wegen @JsExport (ADR-0007); equals/hashCode deshalb von
 * Hand und reihenfolgeunabhängig.
 */
@JsExport
data class Position(
    val board: Array<PlacedPiece>,
    val sideToMove: Color,
    val castlingRights: CastlingRights,
    val enPassantTarget: Square?,
    val halfmoveClock: Int,
    val fullmoveNumber: Int,
) {
    init {
        require(board.map { it.square }.toSet().size == board.size) {
            "Ein Feld ist doppelt besetzt"
        }
    }

    fun pieceAt(square: Square): Piece? = board.firstOrNull { it.square == square }?.piece

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Position) return false
        return pieceMap() == other.pieceMap()
            && sideToMove == other.sideToMove
            && castlingRights == other.castlingRights
            && enPassantTarget == other.enPassantTarget
            && halfmoveClock == other.halfmoveClock
            && fullmoveNumber == other.fullmoveNumber
    }

    override fun hashCode(): Int {
        var r = pieceMap().hashCode()
        r = 31 * r + sideToMove.hashCode()
        r = 31 * r + castlingRights.hashCode()
        r = 31 * r + (enPassantTarget?.hashCode() ?: 0)
        r = 31 * r + halfmoveClock
        r = 31 * r + fullmoveNumber
        return r
    }
}

/** Das Brett als Nachschlagewerk — die interne Sicht der Mechanik. */
internal fun Position.pieceMap(): Map<Square, Piece> = board.associate { it.square to it.piece }

internal fun standardStartPosition(): Position {
    val backRank = listOf(
        PieceType.ROOK, PieceType.KNIGHT, PieceType.BISHOP, PieceType.QUEEN,
        PieceType.KING, PieceType.BISHOP, PieceType.KNIGHT, PieceType.ROOK,
    )
    val pieces = buildList {
        for (file in File.entries) {
            add(PlacedPiece(Square(file, Rank.EIGHT), Piece(backRank[file.ordinal], Color.BLACK)))
            add(PlacedPiece(Square(file, Rank.SEVEN), Piece(PieceType.PAWN, Color.BLACK)))
            add(PlacedPiece(Square(file, Rank.TWO), Piece(PieceType.PAWN, Color.WHITE)))
            add(PlacedPiece(Square(file, Rank.ONE), Piece(backRank[file.ordinal], Color.WHITE)))
        }
    }
    return Position(pieces.toTypedArray(), Color.WHITE, CastlingRights.ALL, null, 0, 1)
}
